src/Parser/TypeParser.php

```diff
@@ -484,7 +492,65 @@ private function parseCallable(TokenIterator $tokens, Ast\Type\IdentifierTypeNod
 		$startIndex = $tokens->currentTokenIndex();
 		$returnType = $this->enrichWithAttributes($tokens, $this->parseCallableReturnType($tokens), $startLine, $startIndex);
 
-		return new Ast\Type\CallableTypeNode($identifier, $parameters, $returnType);
+		return new Ast\Type\CallableTypeNode($identifier, $parameters, $returnType, $templates);
+	}
+
+
+	/**
+	 * @return Ast\Type\CallableTypeTemplateNode[]
+	 *
+	 * @phpstan-impure
+	 */
+	private function parseCallableTemplates(TokenIterator $tokens): array
+	{
+		$tokens->consumeTokenType(Lexer::TOKEN_OPEN_ANGLE_BRACKET);
+
+		$templates = [];
+
+		$isFirst = true;
+		while ($isFirst || $tokens->tryConsumeTokenType(Lexer::TOKEN_COMMA)) {
+			$tokens->tryConsumeTokenType(Lexer::TOKEN_PHPDOC_EOL);
+
+			// trailing comma case
+			if (!$isFirst && $tokens->isCurrentTokenType(Lexer::TOKEN_CLOSE_ANGLE_BRACKET)) {
```

Comment:
I don't understand why is there so much repeated code. We're already parsing template types in similar position for `@method`, there has to be a way to reuse that code: https://github.com/phpstan/phpdoc-parser/pull/160

Everything you should need is already in parseTemplateTagValue.

Reply:
Can you please guide me how and what to reuse? The code in #160 (Parser/PhpDocParser in general) is for phpdoc instead of for type (Parser/TypeParser). If the newly introduced Ast/Type/CallableTypeTemplateNode class is wanted (and it probably is), then the code cannot be simplified much. If you want, feel free to push any changes to my branch.

Reply:
I think #160 implements the very same use case as this PR: it parses inline template tag definitions, disallowing descriptions, but accounting for everything else (bound types, defaults, ...). Perhaps it could suffice to extract the `parseTemplateTagValue` method into a `TemplateTagParser` that could be used in `TypeParser` as well?

Reply:
Yes, that logic should be extracted so that it's usable by both TypeParser and PhpDocParser.

Reply:
@mvorisek are you going to look into this feedback or should i have a go at it?

Reply:
@mad-briller Please have a go at it :)